src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -121,14 +122,16 @@ public int GetHashCode(PSTypeName obj)
         }
     }
 
-    internal class TypeInferenceContext
+    internal class TypeInferenceContext : IDisposable
     {
         public static readonly PSTypeName[] EmptyPSTypeNameArray = Utils.EmptyArray<PSTypeName>();
+        private readonly bool _ownsPowerShell;
         private readonly PowerShell _powerShell;
 
         public TypeInferenceContext()
         : this(PowerShell.Create(RunspaceMode.CurrentRunspace))
         {
+            _ownsPowerShell = true;
```

Comment:
This doesn't look right to me.  This class can own PowerShell/Runspace only if PowerShell is created with `Runspace.NewRunspace` mode.  Otherwise you are disposing a runspace created by someone else (and possibly the thread default runspace).

Reply:
Reverting the whole file.